### ISSUE TITLE
[Feature Request]: Implement Google AI Support custom end point

### ISSUE THREAD

**Issue by H0llyW00dzZ** [ChatGPTNextWeb/ChatGPT-Next-Web#4086]
### Problem Description

In the current version, the service provider `Google` does not support custom endpoints, unlike the service provider `OpenAI`, which provides a custom endpoint URL.

### Solution Description

Implementing support for custom endpoints in the service provider `Google` could greatly enhance performance. As an example, I've been operating this repository on `Kubernetes`, utilizing containers such as [`one-api`](https://github.com/songquanpeng/one-api/pkgs/container/one-api) and [`litellm`](https://github.com/BerriAI/litellm/pkgs/container/litellm) across `13 pods`. This configuration has shown to be more stable compared to deployments on `Vercel` or other platforms.

**Comment by H0llyW00dzZ** [comment]
# Additional Note

> [!NOTE]  
> This improvement could significantly enhance performance not only for web-based applications, but also for desktop versions.

**Comment by fred-bf** [comment]
OneAPI doesnt support Google Gemini API definition, you can diectly config the enpoint in custom api. As for model providers, the progress will be tracked in issue https://github.com/ChatGPTNextWeb/ChatGPT-Next-Web/issues/4030

**Comment by H0llyW00dzZ** [comment]
> OneAPI doesnt support Google Gemini API definition, you can diectly config the enpoint in custom endpoint(GPT like). As for model providers, the progress will be tracked in issue #4030

it's supported

![image](https://github.com/ChatGPTNextWeb/ChatGPT-Next-Web/assets/17626300/88d915cb-8ee0-4d23-893e-61755a44b653)

I've been testing it running on `kubernetes` with better configuration

```sh
(base) root@H0llyW00dzZ:/# kubectl top pods
NAME                       CPU(cores)   MEMORY(bytes)
mysql-78mv3xmb37-hpmz5     5m           402Mi
one-api-j3laxb3o30-bz0gu   0m           46Mi
```

To be honest, I like the `one-api` RESTful APIs because they are written in `Go`, and they also utilize `Gin`.

**Comment by fred-bf** [comment]
@H0llyW00dzZ the base url of gemini api should be able to configured via `GOOGLE_URL`, have you encountered any issue while requesting with the variable?

**Comment by H0llyW00dzZ** [comment]
The reason it can't be used in the desktop version is because it needs to be recompiled again into binaries.

**Comment by H0llyW00dzZ** [comment]
also another reason, rust is so slow when compiling it

**Comment by fred-bf** [comment]
> GOOGLE_URL



> > @H0llyW00dzZ Cool! The base url of gemini api should be able to configured via `GOOGLE_URL`, have you encountered any issue while requesting with the variable?
> 
> no, but env variable GOOGLE_URL can't be used in desktop version
> 
> for example in this one:
> 
> ![image](https://private-user-images.githubusercontent.com/17626300/306705071-de0aa3e7-3447-4fa8-abd4-63b4cb5ebe4b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkwMjY4NTEsIm5iZiI6MTcwOTAyNjU1MSwicGF0aCI6Ii8xNzYyNjMwMC8zMDY3MDUwNzEtZGUwYWEzZTctMzQ0Ny00ZmE4LWFiZDQtNjNiNGNiNWViZTRiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjI3VDA5MzU1MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVkODQ4M2QwMzdhYzhiODhhOGZhYjE3NTFhNmRmZGQwODQzZjU2NTllYjI3NzI4ODRkNjIxNTBjY2ExNGIyNWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.k8Yx5Lj_fNzZPk2BkYxQbSdJdZ0LXxxq9Cm32ouweQE)
> 
> also desktop version it much stable when using custom end point url, plus smoothly unlike web-based applications

They are same thing eventually, you can try switching the endpoint to one-api servoce

**Comment by H0llyW00dzZ** [comment]
> > GOOGLE_URL
> 
> > > @H0llyW00dzZ Cool! The base url of gemini api should be able to configured via `GOOGLE_URL`, have you encountered any issue while requesting with the variable?
> > 
> > 
> > no, but env variable GOOGLE_URL can't be used in desktop version
> > for example in this one:
> > ![image](https://private-user-images.githubusercontent.com/17626300/306705071-de0aa3e7-3447-4fa8-abd4-63b4cb5ebe4b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkwMjY4NTEsIm5iZiI6MTcwOTAyNjU1MSwicGF0aCI6Ii8xNzYyNjMwMC8zMDY3MDUwNzEtZGUwYWEzZTctMzQ0Ny00ZmE4LWFiZDQtNjNiNGNiNWViZTRiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjI3VDA5MzU1MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVkODQ4M2QwMzdhYzhiODhhOGZhYjE3NTFhNmRmZGQwODQzZjU2NTllYjI3NzI4ODRkNjIxNTBjY2ExNGIyNWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.k8Yx5Lj_fNzZPk2BkYxQbSdJdZ0LXxxq9Cm32ouweQE)
> > also desktop version it much stable when using custom end point url, plus smoothly unlike web-based applications
> 
> They are same thing eventually, you can try switching the endpoint to one-api service

No, it doesn't work that way. When using the gemini-pro model, it utilizes Google as a service provider and it's mandatory to use this service.

https://github.com/ChatGPTNextWeb/ChatGPT-Next-Web/blob/main/app/constant.ts#L14C1-L14C77

**Comment by fred-bf** [comment]
@H0llyW00dzZ When requesting gemini api server, the client will check whether the googleUrl is configured first.If not set, will fallback to the url you just mentions.

https://github.com/ChatGPTNextWeb/ChatGPT-Next-Web/blob/08fa22749aea8f497811f684bd9c7ef68d698666/app/api/google/%5B...path%5D/route.ts#L20

**Comment by fred-bf** [comment]
Oh, for native clients,  bypassing a proxy service is still necessary.

**Comment by fred-bf** [comment]
@H0llyW00dzZ have you checked the developer console in development build, it much likely the requests being block by CORS right now.

**Comment by H0llyW00dzZ** [comment]
> @H0llyW00dzZ have you checked the developer console in development build, it much likely the requests being block by CORS right now.

no response

![image](https://github.com/ChatGPTNextWeb/ChatGPT-Next-Web/assets/17626300/cbd3c0c5-cfa4-443f-a8f2-05e4ec325697)

![image](https://github.com/ChatGPTNextWeb/ChatGPT-Next-Web/assets/17626300/4c1847a5-2277-4a72-9776-627663154b01)

![image](https://github.com/ChatGPTNextWeb/ChatGPT-Next-Web/assets/17626300/257180b6-b368-4c20-b710-db83b423afbc)


I am using one-api on k8s

![image](https://github.com/ChatGPTNextWeb/ChatGPT-Next-Web/assets/17626300/8314f707-0607-4477-b0c9-3707f96b381a)

**Comment by H0llyW00dzZ** [comment]
when using service provider openai + model open ai it work, but gemini it doesn't work

**Comment by H0llyW00dzZ** [comment]
in one-api it work while testing

![image](https://github.com/ChatGPTNextWeb/ChatGPT-Next-Web/assets/17626300/cc35b5be-7c34-4f20-a719-f7d3e0e7ebdc)

**Comment by H0llyW00dzZ** [comment]
also I am pretty sure because of this

https://github.com/ChatGPTNextWeb/ChatGPT-Next-Web/blob/08fa22749aea8f497811f684bd9c7ef68d698666/app/client/platforms/google.ts#L121C1-L129C8
